### ISSUE TITLE
Better documention on the chat text format

### ISSUE THREAD

**Issue by michaelnny** [facebookresearch/llama#481]
Hi,

Right now the project only briefly mentions the format for the chat completion in the README.md file.

> The fine-tuned models were trained for dialogue applications. To get the expected features and performance for them, a specific formatting defined in [chat_completion](https://github.com/facebookresearch/llama/blob/main/llama/generation.py#L212) needs to be followed, including the INST and <<SYS>> tags, BOS and EOS tokens, and the whitespaces and breaklines in between (we recommend calling strip() on inputs to avoid double-spaces).

However, the code in https://github.com/facebookresearch/llama/blob/main/llama/generation.py#L212 does not give any examples or explanation on the exact format is being used, and the code is not very clear either.

I think a better documentation on how exactly the prompts are formatted before we apply tokenization might be helpful. At least add some examples would be great.

**Update:**

Here are some examples of the chat text format.

Case 1: Prompt ends at 1st user prompt, not answer yet:

`<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST]`

Case 2: Prompt ends at 2nd user prompt, has 1st answer:

`<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST] {1st response} </s><s>[INST] {2nd user prompt} [/INST]`

Based on these observation, we might create the training data using this format:

```
{"prompt": "<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST]", "completion": " {1st response} </s>"}
{"prompt": "<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST] {1st response} </s><s>[INST] {2nd user prompt} [/INST]", "completion": " {2nd response} </s>"}
```

**Comment by miwojc** [comment]
Is this the format?
```
<<SYS>>
{system_message}
<</SYS>>

[INST] {prompt} [/INST]
```

**Comment by ChaoYan** [comment]
> 

the code from chat completion seems to lead to this format?
```
[INST] <<SYS>>
{sys msg}
<</SYS>>{user msg} [/INST]

**Comment by michaelnny** [comment]
> 



> > 
> 
> The code from chat completion seems to lead to this format?
> 
> ```
> [INST] <<SYS>>
> {sys msg}
> <</SYS>>{user msg} [/INST]
> ```

This seems to be the correct format. 

However I'm curios on why we don't use more general format, like the example mentioned in the OpenAI API document [support-chatbot](https://platform.openai.com/docs/guides/fine-tuning/case-study-customer-support-chatbot)

> "Summary: {summary of the interaction so far}\n\nSystem:{system message}\n\n###\n\nUser: {message1}\nAssistant: {response1}\nUser: {message2}\nAssistant: {response2}\nUser: {message3}\nAssistant:"

Here I made some changes to the above example, so it's more relevant to this conversation. One reason I can think of is we can save few tokens because the OpenAI method potentially has more repeated text (User, Assistant) needs to be encoded, but not sure if this is the correct.

Also, why do we put the system message inside the first instruction block, and not separate it and put it at the beginning of the text?

**Comment by RonanKMcGovern** [comment]
> Also, I just noticed that we're encoding each pair of conversion (user prompt and completion) with bos and eos tokens, instead of tread the full prompt as a single long text, is this really necessary?

If the chat model was trained with turns, then the BOS and EOS possibly contributed to conveying the structure of turns (in combination with INST). Unclear though...

**Comment by RonanKMcGovern** [comment]
I'm also unclear on padding. Using EOS for padding seems to work ok:

```tokenizer.pad_token = tokenizer.eos_token```

I had tried using <pad>:
```
# This seems to give a cuda error
# Add pad token
tokenizer.add_tokens("<pad>")
pad_token_id = tokenizer.get_vocab()["<pad>"]

# Set the pad token in the tokenizer
tokenizer.pad_token = "<pad>"
tokenizer.pad_token_id = pad_token_id

print(f"Pad token: {tokenizer.pad_token}")
print(f"Pad token ID: {tokenizer.pad_token_id}")
```

but this leads to a cuda error, which I think is due to messing up matrix dimensions.

**Comment by michaelnny** [comment]
Here's my attempt to rewrite the build prompt function, the code is much more easy to understand. And we can use it to prepare fine-tuning datasets. Note I haven't run the training script yet as the project is not finished.

```
from typing import Tuple, List, Mapping, Text, Literal, Optional, Tuple, TypedDict

from tokenizer import Tokenizer

Role = Literal["system", "user", "assistant"]


class Message(TypedDict):
    role: Role
    content: str


class CompletionPrediction(TypedDict, total=False):
    generation: str
    tokens: List[str]  # not required
    logprobs: List[float]  # not required


class ChatPrediction(TypedDict, total=False):
    generation: Message
    tokens: List[str]  # not required
    logprobs: List[float]  # not required


Dialog = List[Message]

B_INST, E_INST = "[INST]", "[/INST]"
B_SYS, E_SYS = "<<SYS>>\n", "\n<</SYS>>\n\n"
DEFAULT_SYSTEM_PROMPT = """\
You are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe. Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.

If a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information."""


def maybe_add_system_prompt(dialog: Dialog) -> Dialog:
    """Insert a dialog with system role at the beginning of the list if there're no such one."""
    assert dialog is not None and len(dialog) > 0

    if dialog[0]["role"] != "system":
        dialog = [
            {
                "role": "system",
                "content": DEFAULT_SYSTEM_PROMPT,
            }
        ] + dialog

    return dialog


def build_prompt_completion(
    dialog: Dialog, tokenizer: Tokenizer
) -> Tuple[List[int], List[int]]:
    """Build prompt and completion pair following the Meta llama-2 format.

    Note we only build the training target completion if the last role in the dialog is 'assistant'.

    Here are some examples of the format that llama-2 uses (before apply tokenization), note we inserted BOS and EOS into the example for completeness:
    {"prompt": "<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST]", "completion": " {1st response} </s>"}
    {"prompt": "<s>[INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST] {1st response} </s><s>[INST] {2nd user prompt} [/INST]", "completion": " {2nd response} </s>"}

    """

    assert dialog is not None and len(dialog) > 0

    # make sure the first one is always system prompt
    dialog = maybe_add_system_prompt(dialog)

    assert (
        dialog[0]["role"] == "system"
        and all([msg["role"] == "user" for msg in dialog[1::2]])
        and all([msg["role"] == "assistant" for msg in dialog[2::2]])
    ), (
        "model only supports 'system', 'user' and 'assistant' roles, "
        "starting with 'system', then 'user' and alternating (u/a/u/a/u...)"
    )

    # store user-prompt:answer pairs so we can later add BOS, EOS tokens to each pair
    prompts = []
    for i in range(1, len(dialog), 2):  # skip the first one since it's system prompt
        full_prompt = ""
        prompt = dialog[i]

        if i == 1:
            # add system prompt, note Meta llama-2 insert the system prompt inside the first user prompt
            # as in this format: [INST] <<SYS>>\n{system prompt}\n<</SYS>>\n\n{1st user prompt} [/INST]
            sys_prompt = B_SYS + dialog[0]["content"] + E_SYS
            full_prompt = f"{B_INST} {sys_prompt}{(prompt['content']).strip()} {E_INST}"
        else:
            full_prompt = f"{B_INST} {(prompt['content']).strip()} {E_INST}"

        # add answer to the full prompt
        # here we skip the last answer by the assistant, since it's used for building the training target
        if i + 1 < len(dialog) - 1 and dialog[i + 1]["role"] == "assistant":
            answer = dialog[i + 1]
            full_prompt += f" {(answer['content']).strip()} "

        prompts.append(full_prompt)

    prompt_tokens = []

    # Note llama-2, we add BOS and EOS for every pair of user-prompt:answer, except for the last user-prompt
    for i, prompt in enumerate(prompts):
        if i < len(prompts) - 1:
            tokens = tokenizer.encode(prompt, bos=True, eos=True)
        else:
            tokens = tokenizer.encode(prompt, bos=True, eos=False)
        prompt_tokens.extend(tokens)

    # completion tokens for training
    completion_tokens = None
    if dialog[-1]["role"] == "assistant":
        answer = dialog[-1]
        target = f" {(answer['content']).strip()} "
        completion_tokens = tokenizer.encode(target, bos=False, eos=True)

    return prompt_tokens, completion_tokens


if __name__ == "__main__":
    tokenizer = Tokenizer(model_path="/home/michael/llama-2/tokenizer.model")

    example_dialogs = [
        [
            {"role": "user", "content": "Solve 1+37."},
            {"role": "assistant", "content": "38"},
        ],
        [
            {"role": "user", "content": "Tell me a joke about a dog."},
        ],
        [
            {
                "role": "system",
                "content": "You are a very clever and funny agent, make people laugh is your natural job.",
            },
            {
                "role": "user",
                "content": "Tell me a joke about a cat playing some toy car.",
            },
        ],
        [
            {"role": "user", "content": "I am going to Paris, what should I see?"},
            {"role": "assistant", "content": "You should go to The Eiffel Tower."},
            {"role": "user", "content": "What's so special about it?"},
        ],
        [
            {"role": "user", "content": "I am going to Paris, what should I see?"},
            {"role": "assistant", "content": "You should go to The Eiffel Tower."},
            {"role": "user", "content": "What's so special about it?"},
            {"role": "assistant", "content": "Just go there and you'll find out."},
        ],
    ]

    for dialog in example_dialogs:
        prompt_tokens, completion_tokens = build_prompt_completion(
            dialog,
            tokenizer,
        )

        print(f"Prompt: {tokenizer.decode(prompt_tokens)}")

        if completion_tokens is not None:
            print(f"Completion: {tokenizer.decode(completion_tokens)}")
        print("\n\n")


```

**Comment by michaelnny** [comment]
> I'm also unclear on padding. Using EOS for padding seems to work ok:

I found that the pad id is -1 from Meta's tokenizer (not using Transformers). And if I use -1 to pad the input sequence, the model will threw error during training, and the error is coming from the token embedding layer.

After changing the embedding layer to add padding_idx=-1, it seems to work, however I'm running CUDA out of RAM error, so can't tell it if really works.

**Comment by RonanKMcGovern** [comment]
> If you try to add a new token, is that going to increase the vocab size? Maybe you also need to adjust that, but I'm not sure as I've never done that before.

using <pad> assigns an id of 32000 to it, which I assume is already in the vocab (which then maybe is silly to use as a pad token).

Did you try just using the EOS token to pad?

Let me know if you get -1 working

**Comment by michaelnny** [comment]
> using assigns an id of 32000 to it, which I assume is already in the vocab (which then maybe is silly to use as a pad token).

I think the minimum token id starts from 0, so with a vocabulary size of 32000, the maximum token id should be 32000-1, otherwise you will run into problem as the input sequence contains id greater than the maximum token id, since the embedding layer can't handle that.

Anyway using EOS as the pad id for input sequence shouldn't be a problem, as long as you're padding the target sequence correctly.

**Comment by RonanKMcGovern** [comment]
> I think the minimum token id starts from 0, so with a vocabulary size of 32000, the maximum token id should be 32000-1, otherwise you will run into problem as the input sequence contains id greater than the maximum token id, since the embedding layer can't handle that.

oops, yeah I had assumed vocab was 54k, but you're right it's 32k. Thanks for that.

btw, when you say "target", if I'm doing autoregression, do you mean labels? I'm currently cloning those from input_ids, so that would clone the (wrong) padding token).

I see in your gpt2 (nice!) that you're using the mask to address that? but, as you comment, that increases memory requirements because of precomputing the mask?

**Comment by michaelnny** [comment]
> btw, when you say "target", if I'm doing autoregression, do you mean labels? I'm currently cloning those from input_ids, so that would clone the (wrong) padding token).
> 
> I see in your gpt2 (nice!) that you're using the mask to address that? but, as you comment, that increases memory requirements because of precomputing the mask?

Yes, I'm referring to the labels when I say "target", sorry for the confusion.

In my GPT2 project, I use loss mask because I also want to put less weights on the prompts when computing the loss, which is similar to what OpenAI does. Actually the loss mask does not increase memory that much, it's the attention mask. As the current flash attention kernel implemented in `torch.nn.functional.scaled_dot_product_attention` can't handle both attn_mask and is_causal=True, when use pre-computed attention mask, it will use a much slower kernel branch or something, thus increases memory usage.
